### ISSUE TITLE
A bug in testing framework if error is raised and assertion of the error passes, it omits other assertions

### ISSUE THREAD

**Issue by shahzadlone** [sourcenetwork/defradb#1356]
**Describe the problem**
Our testing frame work seems to be not asserting the other result values once an error is encountered and assertion for `that ExpectedError` passes.

**To Reproduce**
Both the below tests pass (`Results` field in each is different):
```
func TestStopsAssertionOnceRaisedErrorIsAsserted1(t *testing.T) {
	test := testUtils.RequestTestCase{
		Description:   "Delete multiple documents that exist sub selection, should give error.",
		Request:       `mutation { delete_user(ids: ["bae-xyz"]) }`,
		Results:       []map[string]any{{"_key": "Doesn't assert the result if raised expected error."}},
		ExpectedError: "Field \"delete_user\" of type \"[user]\" must have a sub selection.",
	}
	executeTestCase(t, test)
}

func TestStopsAssertionOnceRaisedErrorIsAsserted1(t *testing.T) {
	test := testUtils.RequestTestCase{
		Description:   "Delete multiple documents that exist sub selection, should give error.",
		Request:       `mutation { delete_user(ids: ["bae-xyz"]) }`,
		Results:       []map[string]any{},
		ExpectedError: "Field \"delete_user\" of type \"[user]\" must have a sub selection.",
	}
	executeTestCase(t, test)
}
```

**Expected behavior**
The test should assert other fields (like `Results` in this case) in this case, and ensure tests only pass if all testing args are what we expect them to be.

**Comment by AndrewSisley** [comment]
I do see this as by design, any the results of any action returning an error should not be trusted and should instead be ignored.

The example test cases above use the old system, which only allows one expected error scoped to the entire test case.  The new system scopes the error to a specific request, allowing one to error, and subsequent actions to still be handled correctly.  IIRC a few of our tests rely on this.

Suggest we close this issue.

**Comment by shahzadlone** [comment]
> I do see this as by design, any the results of any action returning an error should not be trusted and should instead be ignored.
> 
> The example test cases above use the old system, which only allows one expected error scoped to the entire test case. The new system scopes the error to a specific request, allowing one to error, and subsequent actions to still be handled correctly. IIRC a few of our tests rely on this.
> 
> Suggest we close this issue.

I am not sure I understand this fully, so I tested this on one of the new tests and the tests still pass if the last action has a matching error but incorrect result.
```

**Comment by AndrewSisley** [comment]
> I am not sure I understand this fully, so I tested this on one of the new tests and the tests still pass if the last action has a matching error but incorrect result. It passed with or without 

Yes, but in that case you are expecting both a result and an error from a single action, which makes no sense IMO.  If you want to test something after an error, use another action.

**Comment by shahzadlone** [comment]
> Yes, but in that case you are expecting both a result and an error from a single action, which makes no sense IMO. If you want to test something after an error, use another action.

Seems like we are saying the same thing. It makes no sense to allow inputting the expected result together with the expected error in the same action. i.e. the testing framework should return an error instead of asserting only half of the action args and passing like it does now.

**Comment by AndrewSisley** [comment]
> > Yes, but in that case you are expecting both a result and an error from a single action, which makes no sense IMO. If you want to test something after an error, use another action.
> 
> Seems like we are saying the same thing. It makes no sense to allow inputting the expected result together with the expected error in the same action. i.e. the testing framework should return an error instead of asserting only half of the action args and passing like it does now.

You want to fail the test if both are provided?  Could make sense

**Comment by shahzadlone** [comment]
> You want to fail the test if both are provided? Could make sense

Yes exactly. Tests should pass only when the input and output is all as we expect.

**Comment by AndrewSisley** [comment]
Ah nice - would be safer than silently ignoring it :)
